File: friendly-coins/java/trunk/src/com/thoughtworks/cschuyle/friendlycoins/solvers/HighestFirstSolver.java
package com.thoughtworks.cschuyle.friendlycoins.solvers;

import static com.thoughtworks.cschuyle.util.WrappedIntegerHelpers.*;
import com.thoughtworks.cschuyle.friendlycoins.primitives.Denomination;
import com.thoughtworks.cschuyle.friendlycoins.primitives.Money;
import com.thoughtworks.cschuyle.friendlycoins.collections.CoinSet;
import com.thoughtworks.cschuyle.friendlycoins.collections.DenominationSet;
import com.thoughtworks.cschuyle.friendlycoins.exception.NoSolutionException;
import com.thoughtworks.cschuyle.WrappedInteger;

public class HighestFirstSolver {

    private HighestFirstSolver() {}
    
    public static CoinSet solve( DenominationSet denominations, Money requiredTotal ) {
        if( Denomination.isEmpty( denominations ) ) {
            throw new NoSolutionException( "No solution - empty denominations set" );
        }
        CoinSet solution = new CoinSet();

        while( null != solution && lessThan( solution.total(), requiredTotal ) ) {
            solution = accumulateHighestPossibleValue( denominations, requiredTotal, solution );
        }
        if( null == solution ) {
            throw new NoSolutionException( denominations, requiredTotal );
        }
        final Money solutionTotal = solution.total();
        if( ! requiredTotal.equals( solutionTotal ) ) {
            throw new IllegalStateException( "Internal Error" );
        }
        return solution;
    }

    private static CoinSet accumulateHighestPossibleValue(
            DenominationSet denominations, Money total, CoinSet destinationCoinSet ) {
        final Money targetTotal = destinationCoinSet.total();
        final WrappedInteger addable = minus( total, targetTotal );

        for( Denomination denomination : denominations ) {
            if( ! greaterThan( denomination, addable ) ) {
                destinationCoinSet = CoinSet.createAugmentedCoinSet( destinationCoinSet, denomination );
                return destinationCoinSet;
            }
        }
        return null;
    }
}

File: friendly-coins/C#/FriendlyCoins/FriendlyCoins/HighestFirstSolver.cs
using System.Collections.Generic;
using System.Linq;

namespace FriendlyCoins
{
    public class HighestFirstSolver
    {
        private readonly IEnumerable<int> denominations;

        public HighestFirstSolver(params int[] denominations)
        {
            this.denominations = denominations.ToArray<int>().OrderBy(i => -i);
        }

        public CoinSet solve(int total)
        {
            var coinSet = new CoinSet();
            while (coinSet.Sum < total)
            {
                accumulateHighest(coinSet, total);
            }
            return coinSet;
        }

        private void accumulateHighest(CoinSet set, int total)
        {
            foreach(int denomination in denominations)
            {
                int available = total - set.Sum;
                if (denomination <= available)
                {
                    set.AddCoin(denomination);
                }
            }
        }
    }
}
